src/libs/Trie/index.js

```diff
@@ -0,0 +1,115 @@
+    * @param {TrieNode} node
+    * @param {Boolean} [allowEmptyWords] - empty word doesn't have any char, you shouldn't pass a true value for it because we are disallowing adding an empty word
+    * @returns {void}
+    */
+    add(word, metaData, node = this.root, allowEmptyWords = false) {
+        const newNode = node;
+        if (word.length === 0 && !allowEmptyWords) {
+            throw new Error('Cannot insert empty word into Trie');
+        }
+        if (word.length === 0) {
+            newNode.isLeaf = true;
+            newNode.metaData = metaData;
+            return;
+        }
+        if (!newNode.children[word[0]]) {
+            newNode.children[word[0]] = new TrieNode();
+            return this.add(word.substring(1), metaData, newNode.children[word[0]], true);
+        }
+        return this.add(word.substring(1), metaData, newNode.children[word[0]], true);
+    }
+
+    /**
+    * Check if the word is exist in the Trie.
+    * @param {String} word
+    * @returns {Object|null} – the node for the word if it's found, or null if it's not found
+    */
+    search(word) {
+        let newWord = word;
+        let node = this.root;
+        while (newWord.length > 1) {
+            if (!node.children[newWord[0]]) {
+                return null;
+            }
+            node = node.children[newWord[0]];
+            newWord = newWord.substring(1);
+        }
+        return node.children[newWord] && node.children[newWord].isLeaf ? node.children[newWord] : null;
+    }
+
+    /**
+    * Update a word data in the Trie.
+    * @param {String} word
+    * @param {Object} metaData
+    * @returns {void}
+    */
+    update(word, metaData) {
+        let newWord = word;
+        let node = this.root;
+        while (newWord.length > 1) {
+            if (!node.children[newWord[0]]) {
+                return null;
+            }
+            node = node.children[newWord[0]];
+            newWord = newWord.substring(1);
+        }
+        node.children[newWord].metaData = metaData;
+    }
+
+    /**
+    * Find all leaf nodes starting with a substring.
+    * @param {String} substr
+    * @param {Number} [limit] - matching words limit
+    * @returns {Array}
+    */
+    getAllMatchingWords(substr, limit = 4) {
+        let node = this.root;
+        let prefix = '';
+        for (let i = 0; i < substr.length; i++) {
+            prefix += substr[i];
+            if (node.children[substr[i]]) {
+                node = node.children[substr[i]];
+            } else {
+                return [];
+            }
+        }
+        return this.getChildMatching(node, prefix, [], limit);
+    }
+
+    /**
+    * Find all leaf nodes that are descendants of a given child node.
+    * @param {TrieNode} node
+    * @param {String} prefix
+    * @param {Array} [words]
+    * @param {Number} limit
+    * @returns {Array}
+    */
+    getChildMatching(node, prefix, words = [], limit) {
```

Comment:
It's incorrect to have a required parameter after an optional parameter, I'm surprised this is passing lint. Need to change the order or give `limit` a default (`4`?)

Reply:
Done, we already gave `limit` in `getAllMatchingWords` method and `getChildMatching` is only called by `getAllMatchingWords` method.